src/core/matching_state.rs

```diff
@@ -1,78 +1,114 @@
-use super::{Entry, Heuristic, LangData, MatchData, Sender};
+use super::*;
 use std::{
-    collections::HashMap,
+    any::Any,
     ffi::{OsStr, OsString},
-    path::PathBuf,
+    ops::DerefMut,
 };
 
-struct PathData<'entry> {
-    entry: &'entry mut Entry,
-    weight: i32,
-    reasons: Vec<LangData>,
-}
-
+/// State passed to heuristics to manipulate matches and query current directory contents.
+///
+/// Only this type should be used to interact with the filesystem and return meaningful heuristic result to the user.
 pub struct MatchingState<'entries> {
-    contents: HashMap<OsString, PathData<'entries>>,
+    /// Optimized storage for current direcotry contents.
+    contents: HashMap<OsString, (&'entries mut Entry, Vec<MatchDataBuilder>)>,
```

Comment:
It vec of builders or vector of builded MatchData?

Reply:
Yes, it is a Vec of builders to avoid making a lot of structs with nearly the same fields. Trust me, that way there is a lot less boilerplate and the builders only get combined into one after all heuristics have been run.

Reply:
https://github.com/Kacper0510/project_cleaner/pull/7#discussion_r1727473129